sqlglot/builder.py

```diff
@@ -0,0 +1,74 @@
+from functools import wraps
+
+import sqlglot.expressions as exp
+
+
+def builder(func):
+    @wraps(func)
+    def builder_method(self, *args, **kwargs):
+        query = Query()
+        query.node = self.node.copy()
+
+        rv = func(query, *args, **kwargs)
+        if rv is not None:
+            return rv
+        return query
+
+    return builder_method
+
+
+class Query:
+    def __init__(self):
+        self.node = exp.Select()
+
+    def to_node(self, obj, spec=None, prefix=None):
+        if isinstance(obj, exp.Expression):
+            return obj
+        if isinstance(obj, Query):
+            return sql.node
+
+        return sqlglot.parse_one(
+            f"{prefix} {obj}" if prefix else str(obj),
+            spec=spec,
+        )
+
+    @builder
+    def select(self, *selects):
+        self.node.args["expressions"].extend(self.to_node(select) for select in selects)
+
+    @builder
+    def from_(self, obj):
+        self.node.args["from"] = self._parse_from(obj)
+
+    @builder
+    def group_by(self, expression):
+        self.node.args["group"] = self.to_node(
+            expression, spec=lambda parser: parser._parse_group, prefix="GROUP BY"
```

Comment:
Calling `_parse_*` is a bit hacky. I think some linters should complain about this?

Maybe we can do something similar to this from within the parser class: https://github.com/tobymao/sqlglot/blob/92735b4a754674c61fad3624a634ffc2ab9a2c42/sqlglot/generator.py#L200

The `Expression.key` attribute can be made a class attribute:


```python
class classproperty:
    def __init__(self, f):
        self.f = f

    def __get__(self, obj, owner):
        return self.f(owner)

class Expression:
    ...

    @classproperty
    def key(cls):
        return cls.__name__.lower()
```

Then you can do something like:

```python
class Parser:
    ...
    
    def parse_into(code, expression_type):
        get_attr(f"_parse_{expression_type.key}")(code)

Parser().parse_into("x.y", exp.Table)
```

Reply:
Or just make public methods for each type haha

Reply:
which do you prefer? public or magic method? i can do that refactor in a separate branch

Reply:
Given that there isn't guaranteed to be a method for each expression type, probably safer to have a slew of public methods.